[1/N] [DTensor device order] Device mesh util function to support device order placement

zpcore opened the pull request:
Improve some utils functions in device mesh in order to support device order placement


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165197
* #165092
* #164902
* #164821
* #164820
* #164806
* __->__ #164797



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

zpcore committed:
Device mesh util function to support device order placement

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -1001,11 +1001,33 @@ def ndim(self) -> int:
             return tuple(self.mesh.shape)
 
-        def get_rank(self) -> int:
+        def get_rank(self, coordinate: Optional[list[int]] = None) -> int:
```
hmmm can we not mix it with this API? I think they are doing two different things, why are we doing this piggyback? Also, I am wondering if this function can be inside DTensor.

One more thing, since now device mesh is using layout, any chance you can directly use `crd2idx` inside torch/distributed/_pycute/int_tuple.py? That way you don't really need this code change.

fduwjj reviewed (commented):
I have some questions regarding the necessity of creating new API and the change to the existing API..

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[1/N] [DTensor device order] Device mesh util function to support device order placement"

Improve some utils functions in device mesh in order to support device order placement





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
